system/HTTP/SiteURIFactory.php

```diff
@@ -0,0 +1,245 @@
+     * fixing the query string if necessary.
+     *
+     * This method updates superglobal $_SERVER and $_GET.
+     *
+     * @return string The route path (before normalization).
+     */
+    private function parseRequestURI(): string
+    {
+        if (
+            $this->superglobals->server('REQUEST_URI') === null
+            || $this->superglobals->server('SCRIPT_NAME') === null
+        ) {
+            return '';
+        }
+
+        // parse_url() returns false if no host is present, but the path or query
+        // string contains a colon followed by a number. So we attach a dummy
+        // host since REQUEST_URI does not include the host. This allows us to
+        // parse out the query string and path.
+        $parts = parse_url('http://dummy' . $this->superglobals->server('REQUEST_URI'));
+        $query = $parts['query'] ?? '';
+        $path  = $parts['path'] ?? '';
+
+        // Strip the SCRIPT_NAME path from the URI
+        if (
+            $path !== '' && isset($this->superglobals->server('SCRIPT_NAME')[0])
+            && pathinfo($this->superglobals->server('SCRIPT_NAME'), PATHINFO_EXTENSION) === 'php'
+        ) {
+            // Compare each segment, dropping them until there is no match
+            $segments = $keep = explode('/', $path);
+
+            foreach (explode('/', $this->superglobals->server('SCRIPT_NAME')) as $i => $segment) {
+                // If these segments are not the same then we're done
+                if (! isset($segments[$i]) || $segment !== $segments[$i]) {
+                    break;
+                }
+
+                array_shift($keep);
+            }
+
+            $path = implode('/', $keep);
+        }
+
+        // This section ensures that even on servers that require the URI to
+        // contain the query string (Nginx) a correct URI is found, and also
+        // fixes the QUERY_STRING Server var and $_GET array.
+        if (trim($path, '/') === '' && strncmp($query, '/', 1) === 0) {
+            $parts    = explode('?', $query, 2);
+            $path     = $parts[0];
+            $newQuery = $query[1] ?? '';
+
+            $this->superglobals->setServer('QUERY_STRING', $newQuery);
+        } else {
+            $this->superglobals->setServer('QUERY_STRING', $query);
+        }
+
+        // Update our global GET for values likely to have been changed
+        parse_str($this->superglobals->server('QUERY_STRING'), $get);
+        $this->superglobals->setGetArray($get);
+
+        return URI::removeDotSegments($path);
+    }
+
+    /**
+     * Will parse QUERY_STRING and automatically detect the URI from it.
+     *
+     * This method updates superglobal $_SERVER and $_GET.
+     *
+     * @return string The route path (before normalization).
+     */
+    private function parseQueryString(): string
+    {
+        $query = $this->superglobals->server('QUERY_STRING') ?? @getenv('QUERY_STRING');
```

Comment:
What are the potential errors suppressed here? 

> Returns the value of the environment variable name, or false if the environment variable name does not exist.

I think we need to cast to string if we are going to trim this below.

Reply:
I don't know. It came from here:
https://github.com/codeigniter4/CodeIgniter4/commit/a8c5ac5df07#diff-2f873856503e4e4cdeca5087aa04d9944d5d732a92a1f1d756b5288a2b4b582dR402

Reply:
Refactored. 78b0ac6ef9d1fed2a48d16d74b1389a550146d7e

Reply:
Thanks! I think that is better.